### PR TITLE
Enable mismatched data size check only when HAVE_ERROR_CHECKING is defined

### PR DESCRIPTION
**Pull request by jain-surabhi-23**
Checking of amount of data received and expected should be done only
when HAVE_ERROR_CHECKING is defined

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by jain-surabhi-23** [comment]
@raffenet Can you please review this PR?

File: src/mpi/coll/bcast/bcast_intra_binomial.c [hunk 1]
```diff
@@ -22,8 +22,12 @@ int MPIR_Bcast_intra_binomial(void *buffer,
     int mpi_errno = MPI_SUCCESS;
     int mpi_errno_ret = MPI_SUCCESS;
     MPI_Aint nbytes = 0;
+#ifdef HAVE_ERROR_CHECKING
     MPI_Aint recvd_size;
     MPI_Status status;
+#else
+    MPI_Status status = *MPI_STATUS_IGNORE;
```
Comment (raffenet) [inline src/mpi/coll/bcast/bcast_intra_binomial.c]: Surely this will cause a segfault, no?

Comment (jain-surabhi-23) [reply src/mpi/coll/bcast/bcast_intra_binomial.c]: Ah. yes. Sorry. I overlooked. Fixed it now.

**Comment by jain-surabhi-23** [comment]
@raffenet Is the overall idea correct here in the PR? These kind of error checks can sometimes be very expensive in some collective algorithms, so we want to enable it only when MPICH is configured with "--enable-error-checking=yes". 

In the case of "optimized" builds we can configure with "--enable-error-checking=no" and performance will not be hurt.

**Comment by raffenet** [comment]
Yes, I believe this is the right approach. Detecting mismatch is not necessary for correct programs, therefore there checks are optional.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
Can you do the same for other bcast impls for consistency? Looks like there's 4 total.
```
bcast_intra_binomial.c:                              "**collective_size_mismatch",
bcast_intra_binomial.c:                              "**collective_size_mismatch %d %d", recvd_size, nbytes);
bcast_intra_scatter_recursive_doubling_allgather.c:                      "**collective_size_mismatch",
bcast_intra_scatter_recursive_doubling_allgather.c:                      "**collective_size_mismatch %d %d", curr_size, nbytes);
bcast_intra_scatter_ring_allgather.c:                      "**collective_size_mismatch",
bcast_intra_scatter_ring_allgather.c:                      "**collective_size_mismatch %d %d", curr_size, nbytes);
bcast_intra_smp.c:                                  "**collective_size_mismatch",
bcast_intra_smp.c:                                  "**collective_size_mismatch %d %d", recvd_size, nbytes);
```

**Comment by jain-surabhi-23** [comment]
@raffenet The requested changes have been made.

**Comment by hzhou** [comment]
@jain-surabhi-23 There is a outstanding test failure probably related:
```
not ok 1506 - ./errors/coll/bcastlength 4
  ---
  Directory: ./errors/coll
  File: bcastlength
  Num-procs: 4
  Timeout: 180
  Date: "Wed Sep 11 19:35:13 2019"
  ...
## Test output (expected 'No Errors'):
## Did not detect mismatched length (long) on process 3
## Did not detect mismatched length (long) on process 2
## Did not detect mismatched length (short) on process 2
## Did not detect mismatched length (short) on process 3
##  Found 4 errors
```
This is in `ofi` config when `izem` is configured. Seems the `release_gather` version of `bcast` failed to do the same check.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Pull request merged**